src/inmanta/server/services/environment_metrics_service.py

```diff
@@ -0,0 +1,199 @@
+"""
+    Copyright 2022 Inmanta
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+import abc
+import logging
+from collections.abc import Sequence
+from datetime import datetime, timedelta
+from enum import Enum
+from typing import Dict, List, Optional
+
+import asyncpg
+
+from inmanta.data import EnvironmentMetricsGauge, EnvironmentMetricsTimer
+from inmanta.server import SLICE_DATABASE, SLICE_ENVIRONMENT_METRICS, SLICE_TRANSPORT, protocol
+
+LOGGER = logging.getLogger(__name__)
+
+COLLECTION_INTERVAL_IN_SEC = 60
+METRIC_NAME_SEPARATOR = "."
+
+
+class MetricType(str, Enum):
+    """
+    There are 3 types of metrics: gauge, timer and meter metrics.
+
+    gauge: will do instantaneous readings of particular values at timestamps
+    timer: will count the number of time events occurred and the total time the events took.
+    meter: will measure the rate of events over time.
+    """
+
+    GAUGE = "gauge"
+    TIMER = "timer"
+    METER = "meter"
+
+
+class MetricValue:
+    """
+    the Metric values as they should be returned by a MetricsCollector of type gauge
+
+    The name of a metric as stored in the DB is the concatenation of metric_name and a field on which the data is grouped by.
+    The METRIC_NAME_SEPARATOR '.' is used as separator and should not be used in the name itself.
+    For example, a metric collected by for the agent_count collector and grouped for the state "up",
+    will have agent_count.up as metric.
+    if the metric is not grouped by anything the name of the collector is used as metric_name
+    """
+
+    def __init__(self, metric_name: str, count: int, grouped_by: Optional[str] = None) -> None:
+        if METRIC_NAME_SEPARATOR in metric_name:
+            raise Exception(
+                'The value "%s" can not be used in the metric_name (%s) as it is used as separator'
+                % (METRIC_NAME_SEPARATOR, metric_name)
+            )
+        if grouped_by and METRIC_NAME_SEPARATOR in grouped_by:
+            raise Exception(
+                'The value "%s" can not be used in the grouped_by value (%s) as it is used as separator'
+                % (METRIC_NAME_SEPARATOR, grouped_by)
+            )
```

Comment:
Very much a detail...
```suggestion
            raise Exception(
                'The character "%s" can not be used in the metric_name (%s)'
                % (METRIC_NAME_SEPARATOR, metric_name)
            )
        if grouped_by and METRIC_NAME_SEPARATOR in grouped_by:
            raise Exception(
                'The character "%s" can not be used in the grouped_by value (%s)'
                % (METRIC_NAME_SEPARATOR, grouped_by)
            )
```